keras/applications/densenet.py

```diff
@@ -0,0 +1,333 @@
+                filename = 'densenet169_weights_tf_dim_ordering_tf_kernels.h5'
+                weights_path = get_file(
+                    filename, __url_prefix__ + filename,
+                    cache_subdir='models',
+                    file_hash='bcf9965cf5064a5f9eb6d7dc69386f43')
+            elif blocks == [6, 12, 48, 32]:
+                filename = 'densenet201_weights_tf_dim_ordering_tf_kernels.h5'
+                weights_path = get_file(
+                    filename, __url_prefix__ + filename,
+                    cache_subdir='models',
+                    file_hash='7bb75edd58cb43163be7e0005fbe95ef')
+        else:
+            if blocks == [6, 12, 24, 16]:
+                filename = 'densenet121_weights_tf_dim_ordering_tf_kernels_notop.h5'
+                weights_path = get_file(
+                    filename, __url_prefix__ + filename,
+                    cache_subdir='models',
+                    file_hash='4912a53fbd2a69346e7f2c0b5ec8c6d3')
+            elif blocks == [6, 12, 32, 32]:
+                filename = 'densenet169_weights_tf_dim_ordering_tf_kernels_notop.h5'
+                weights_path = get_file(
+                    filename, __url_prefix__ + filename,
+                    cache_subdir='models',
+                    file_hash='50662582284e4cf834ce40ab4dfa58c6')
+            elif blocks == [6, 12, 48, 32]:
+                filename = 'densenet201_weights_tf_dim_ordering_tf_kernels_notop.h5'
+                weights_path = get_file(
+                    filename, __url_prefix__ + filename,
+                    cache_subdir='models',
+                    file_hash='1c2de60ee40562448dbac34a0737e798')
+        model.load_weights(weights_path)
+    elif weights is not None:
+        model.load_weights(weights)
+
+    return model
+
+
+def DenseNet121(include_top=True,
+                weights='imagenet',
+                input_tensor=None,
+                input_shape=None,
+                pooling=None,
+                classes=1000):
+    return DenseNet([6, 12, 24, 16],
+                    include_top, weights,
+                    input_tensor, input_shape,
+                    pooling, classes)
+
+
+def DenseNet169(include_top=True,
+                weights='imagenet',
+                input_tensor=None,
+                input_shape=None,
+                pooling=None,
+                classes=1000):
+    return DenseNet([6, 12, 32, 32],
+                    include_top, weights,
+                    input_tensor, input_shape,
+                    pooling, classes)
+
+
+def DenseNet201(include_top=True,
+                weights='imagenet',
+                input_tensor=None,
+                input_shape=None,
+                pooling=None,
+                classes=1000):
+    return DenseNet([6, 12, 48, 32],
+                    include_top, weights,
+                    input_tensor, input_shape,
+                    pooling, classes)
+
+
+def preprocess_input(x):
```

Comment:
Is this a different preprocessing mode than what we already cover in `imagenet_utils`?

Note that `preprocess_input` should be implemented to work with both Numpy arrays and symbolic tensors. It should also be efficient (no memory copy).

Reply:
Yes, I borrowed the preprocessing method from \`https://github.com/liuzhuang13/DenseNet\` which is the same as \`https://github.com/facebook/fb.resnet.torch\`. I think that this function is hard to be integrated into \`imagenet_utils\` because it includes broadcasted division which is not implemented in the current version of the three backends.

Reply:
Yes, I borrowed the preprocessing method from `https://github.com/liuzhuang13/DenseNet` which is the same as `https://github.com/facebook/fb.resnet.torch`. I think that this function is hard to be integrated into `imagenet_utils` because it includes broadcasted division which is not implemented in the current version of the three backends.